### PR TITLE
Create custom exceptions for spark utilities 

### PR DESCRIPTION
**Pull request by vansika**
I have been thinking about it for some time. Though all the possible exceptions have been taken care of, I very deeply feel that there is duplication of code and that it looks ugly. Now that most of the things are clear, and  I am working on the recently added schema (model_metadata) in a separate branch, I strongly feel about making the exception handling stuff prettier. 

This PR just reflects my thinking process (that is why does not include changes at all the places) and is solely for the purpose of receiving suggestions and explanations on how errors can be taken care of. We can certainly close this PR in time and open a new one (or include all the changes here itself), as it suits us.

### CODE REVIEW

**Comment by vansika** [comment]
Reviews requested from @mayhem @alastair @paramsingh.
If anyone else wants to review, you're most welcome.

**Comment by paramsingh** [comment]
This looks really good!

**Comment by vansika** [comment]
I think this handles exceptions in a good manner (still minimal). These changes are good to move forward with the project. In time, we will, of course, improve a lot on exception handling.

**Comment by vansika** [comment]
A thumbs up and then I can reflect the changes in the other scripts of our basic recommendation engine and then it is good to merge.

File: requirements.txt [hunk 1]
```diff
@@ -5,3 +5,4 @@ Jinja2 == 2.10.1
 pika==0.13.0
 python-dateutil==2.8.0
 numpy==1.17.2
+git+https://github.com/metabrainz/brainzutils-python.git@v1.10.1
```
Comment (paramsingh) [inline requirements.txt]: Is there a newer version we could use?

File: manage.py [hunk 2]
```diff
@@ -76,28 +78,32 @@ def dataframes():
...
 def recommend():
     """ Invoke script responsible for generating recommendations.
     """
     from listenbrainz_spark.recommendations import recommend
-    recommend.main()
+    with app.app_context():
+        recommend.main()
 
 @cli.command(name='user')
 def user():
```
Comment (paramsingh) [inline manage.py]: Why not log this to sf try as well?

**Comment by vansika** [comment]
candidate_sets and recommend are yet to be changed. (after the exam positively) :(

File: listenbrainz_spark/recommendations/train_models.py [hunk 3]
```diff
@@ -86,18 +88,20 @@ def train(training_data, validation_data, num_validation, ranks, lambdas, iterat
     alpha = 3.0
     for rank, lmbda, iteration in itertools.product(ranks, lambdas, iterations):
         t0 = time()
+        model_id = 'listenbrainz-recommendation-model-{}'.format(uuid.uuid4())
         try:
             model = ALS.trainImplicit(training_data, rank, iterations=iteration, lambda_=lmbda, alpha=alpha)
-            model_id = 'listenbrainz-recommendation-model-{}'.format(uuid.uuid4())
         except Py4JJavaError as err:
-            raise Py4JJavaError('Unable to train model "{}": {}\n'.format(model_id, type(err).__name__), err.java_exception)
+            current_app.logger.error('Unable to train model "{}": {}\n{}'.format(model_id, type(err).__name__, str(err.java_exception)))
```
Comment (paramsingh) [inline listenbrainz_spark/recommendations/train_models.py]: This should have exc_info set to true right?

File: listenbrainz_spark/recommendations/train_models.py [hunk 4]
```diff
@@ -86,18 +88,20 @@ def train(training_data, validation_data, num_validation, ranks, lambdas, iterat
...
+            sys.exit(-1)
         mt = '{:.2f}'.format((time() - t0) / 60)
         t0 = time()
         try:
             validation_rmse = compute_rmse(model, validation_data, num_validation)
         except Py4JJavaError as err:
-            raise Py4JJavaError('Root Mean Squared Error for model "{}" for validation data not computed: {}\n'.format(
-                model_id, type(err).__name__), err.java_exception)
+            current_app.logger.error('Root Mean Squared Error for model "{}" for validation data not computed: {}\n{}'.format(
+                model_id, type(err).__name__, str(err.java_exception)))
```
Comment (paramsingh) [inline listenbrainz_spark/recommendations/train_models.py]: So should this.

File: listenbrainz_spark/recommendations/train_models.py [hunk 5]
```diff
@@ -174,34 +175,31 @@ def main():
...
-    logging.info('Saving model...')
+    current_app.logger.info('Saving model...')
     t0 = time()
+    metadata_file_path = os.path.join(path.DATA_DIR, best_model_metadata['model_id'])
     try:
-        metadata_file_path = os.path.join(path.DATA_DIR, best_model_metadata['model_id'])
         model.model.save(listenbrainz_spark.context, config.HDFS_CLUSTER_URI + metadata_file_path)
     except Py4JJavaError as err:
-        logging.error('Unable to save best model "{}": {} \n{}. Aborting...'.format(best_model_metadata['model_id'],
+        current_app.logger.error('Unable to save best model "{}": {} \n{}. Aborting...'.format(best_model_metadata['model_id'],
```
Comment (paramsingh) [inline listenbrainz_spark/recommendations/train_models.py]: And this.

**Pull request merged**